syz-manager/rpc.go

```diff
@@ -366,3 +364,147 @@ func (runner *Runner) newRequest(req *fuzzer.Request) rpctype.ExecutionRequest {
 		NeedHints:    req.NeedHints,
 	}
 }
+
+type requestDistributor struct {
+	fuzzer           *fuzzer.Fuzzer
+	statReusedSafe   *stats.Val
+	statReusedUnsafe *stats.Val
+	// These are the inputs that were sent to VMs, but were not executed.
+	safeInputs   *queue[fuzzer.Request] // from stopped VMs
+	unsafeInputs *queue[fuzzer.Request] // from crashed VMs
+
+	// We don't want to rescue unsafe inputs twice.
+	rescuedUnsafe sync.Map
+}
+
+func newRequestDistributor(fuzzerObj *fuzzer.Fuzzer, statReusedSafe, statReusedUnsafe *stats.Val) *requestDistributor {
+	return &requestDistributor{
+		fuzzer:           fuzzerObj,
+		safeInputs:       newQueue[fuzzer.Request](1000),
+		unsafeInputs:     newQueue[fuzzer.Request](1000),
+		statReusedSafe:   statReusedSafe,
+		statReusedUnsafe: statReusedUnsafe,
+	}
+}
+
+func (rd *requestDistributor) Next(runner *Runner) *fuzzer.Request {
+	if rd.mayUnsafe(runner) {
+		req := rd.unsafeInputs.Fetch()
+		if req != nil {
+			rd.statReusedUnsafe.Add(1)
+			rd.rescuedUnsafe.Store(req, true)
+			return req
+		}
+	}
+	// Next() must be as fast as possible, so let's avoid contention.
+	req := rd.safeInputs.TryFetch()
+	if req != nil {
+		rd.statReusedSafe.Add(1)
+		return req
+	}
+	return rd.fuzzer.NextInput()
+}
+
+func (rd *requestDistributor) mayUnsafe(runner *Runner) bool {
+	if !runner.mu.TryLock() {
+		return false
+	}
+	defer runner.mu.Unlock()
+	uptime := time.Since(runner.started)
+
+	// We don't want to risk crashing freshly booted VMs.
+	// So let's limit the time wasted on VM re-creation to 10% of the total uptime.
+	if (runner.bootTime+uptime).Seconds()*0.1 < runner.bootTime.Seconds() {
+		return false
+	}
+
+	const unsafeOnceIn = time.Minute
+	if time.Since(runner.lastUnsafe) < unsafeOnceIn {
```

Comment:
Giving 1 candidate per exchange is not enough?
The fuzzer queue is 2*procs, so unsafe progs should be ~2*procs apart.

Reply:
If exchanges are fast (at least in a local qemu-based setup it is so), it sends 1-2 new programs at a time. If we include an unsafe request every time, that would be too frequent.